src/sentry/web/frontend/accounts.py

```diff
@@ -373,44 +381,42 @@ def disconnect_identity(request, identity_id):
 @login_required
 def show_emails(request):
     user = request.user
+    emails = user.emails.all()
+    email_form = EmailForm(user, request.POST or None)
     primary_email = UserEmail.get_primary_email(user)
-    alt_emails = user.emails.all().exclude(email=primary_email.email)
-
-    email_form = EmailForm(user, request.POST or None,
-        initial={
-            'primary_email': primary_email.email,
-        },
-    )
+    alt_emails = emails.exclude(email=primary_email.email)
 
     if 'remove' in request.POST:
         email = request.POST.get('email')
         del_email = UserEmail.objects.filter(user=user, email=email)
         del_email.delete()
         return HttpResponseRedirect(request.path)
 
-    if email_form.is_valid():
-        old_email = user.email
+    if 'primary' in request.POST:
+        new_primary = request.POST.get('new_primary_email')
+        if new_primary != user.email:
 
-        email_form.save()
+            # update notification settings for those set to primary email with new primary email
+            alert_email = UserOption.objects.get_value(user=user, project=None, key='alert_email')
 
-        if user.email != old_email:
-            useroptions = UserOption.objects.filter(user=user, value=old_email)
-            for option in useroptions:
+            if alert_email == user.email:
+                UserOption.objects.set_value(user=user, project=None, key='alert_email', value=new_primary)
+            options = UserOption.objects.filter(user=user, key='mail:email')
+            for option in [o for o in options if o.value == user.email]:
                 option.value = user.email
                 option.save()
-            UserEmail.objects.filter(user=user, email=old_email).delete()
-            try:
-                with transaction.atomic():
-                    user_email = UserEmail.objects.create(
-                        user=user,
-                        email=user.email,
-                    )
-            except IntegrityError:
-                pass
-            else:
-                user_email.set_hash()
-                user_email.save()
-            user.send_confirm_emails()
+
+            new_username = user.email == user.username
+
+            user.email = new_primary
+
+            if new_username and not User.objects.filter(username__iexact=new_primary).exists():
```

Comment:
Hmm, this scares me a bit.

If there is a `User` who already has this email address as their username, I think we should hard error here. Unless I'm not quite understanding, this seems like a bad situation to be in. I'm also not sure if this is behavior we'd want. But that's more of a product decision since ticking a new primary email address changes how a user would login to their account.

Also, this path isn't quite transactionally safe. It's possible that a row might not exist when this check is made, but then error on the `.save()` below.